### PR TITLE
feat: add decoded response to decodeExpecting errors

### PR DESCRIPTION
**Pull request by bitgopatmcl**
Should address #137 

Also it brings in `PathReporter` which has a bug as per #138, but we were calling `JSON.stringify` on the error anyway so it doesn't introduce a regression. Will try to sort that out in a follow-up.

### CODE REVIEW

File: packages/superagent-wrapper/src/request.ts [hunk 1]
```diff
@@ -148,7 +158,7 @@ const patchRequest = <
     patchedReq.decode().then((res) => {
       if (res.status !== status) {
         const error = res.error ?? `Unexpected status code ${String(res.status)}`;
```
Comment (ericcrosson-bitgo) [inline packages/superagent-wrapper/src/request.ts:160]: `error` perhaps a misleading name for this variable now

**Review by ericcrosson-bitgo — APPROVED** [review]
Looks fine to me, I don't see any breaking changes.

I quite dislike `Error`s and especially subclassing `Error`s, but `decodeExpecting` is designed to `throw` and there's no better value to `throw` than an `Error`.

Lastly, I want to make sure this change identifies the entirety of #137. I see

> The error message would be more helpful if it included a stringified version of the decoded error response

as best I can tell this is the main ask, so looks like this PR hits the :dart:

**Comment by ericcrosson-bitgo** [comment]
Oh, I think I see what you mean. So instead of saying "I tried to decode 200Codec and got these errors", it's sufficient to say "I expected a 200 but got this status code and response instead"?

**Comment by bitgopatmcl** [comment]
Maybe it would be helpful to list out the possibilities and what should happen:

- Expected status code, decoded correctly: nothing thrown, result returned like normal
- Unexpected status code, decoded correctly: error thrown: "Unexpected response 404: { message: "Thing not found" }"
- Expected status code, decode errors: error thrown: "Could not decode response 200: { stuff: [ ... ] }"
- Unexpected status code, decode errors: error thrown: "Unexpected response 418: { message: "I'm a teapot" }"
- Unknown status code: error thrown: "Unexpected response 404: { error: "oops" }"

Do these seem right?

File: packages/superagent-wrapper/test/request.test.ts [hunk 2]
```diff
@@ -169,21 +181,32 @@ describe('request', () => {
...
         .post({ id: 1337, foo: 'test', bar: 42 })
         .set('x-send-invalid-response-body', 'true')
         .decodeExpecting(200)
-        .then(() => false)
-        .catch(() => true);
+        .then(() => '')
+        .catch((err) => (err instanceof DecodeError ? err.message : ''));
 
-      assert.isTrue(result);
+      assert.deepEqual(result, 'Could not decode response 200: {"invalid":"response"}');
```
Comment (ericcrosson-bitgo) [inline packages/superagent-wrapper/test/request.test.ts:209]: `deepEqual` an odd choice here!

File: packages/superagent-wrapper/src/request.ts [hunk 3]
```diff
@@ -146,9 +156,16 @@ const patchRequest = <
...
-        throw new Error(JSON.stringify(error));
+      if (res.original.status !== status) {
+        const error = `Unexpected response ${String(
+          res.original.status,
+        )}: ${JSON.stringify(res.original.body)}`;
+        throw new DecodeError(error, res as DecodedResponse<h.HttpRoute>);
+      } else if (res.status === 'decodeError') {
+        const error = `Could not decode response ${String(
+          res.original.status,
+        )}: ${JSON.stringify(res.original.body)}`;
```
Comment (ericcrosson-bitgo) [inline packages/superagent-wrapper/src/request.ts:167]: Note that `JSON.stringify` can also throw :sob: I believe when a circular reference is encountered.

fp-ts [provides a wrapper](https://gcanti.github.io/fp-ts/modules/Json.ts.html#stringify) that returns an `Either`, which we may want to use

File: packages/superagent-wrapper/src/request.ts [hunk 4]
```diff
@@ -146,9 +156,16 @@ const patchRequest = <
     status: StatusCode,
   ) =>
     patchedReq.decode().then((res) => {
-      if (res.status !== status) {
-        const error = res.error ?? `Unexpected status code ${String(res.status)}`;
-        throw new Error(JSON.stringify(error));
+      if (res.original.status !== status) {
+        const error = `Unexpected response ${String(
+          res.original.status,
+        )}: ${JSON.stringify(res.original.body)}`;
```
Comment (ericcrosson-bitgo) [inline packages/superagent-wrapper/src/request.ts:162]: Is `res.original.status` not already coerced to a string by the backticks?
```suggestion
        const error = `Unexpected response ${res.original.status}: ${JSON.stringify(res.original.body)}`;
```

**Comment by ericcrosson-bitgo** [comment]
> Maybe it would be helpful to list out the possibilities and what should happen:

I hope we have tests for each case, I'll take a look

**Review by ericcrosson-bitgo — APPROVED** [review]
This change looks fine to me, it modifies `decodeExpecting` to return a particular error (`DecodeError` -- note @bitgopatmcl that this may be ambiguous with an io-ts decode error _outside_ of an api-ts context) so we can identify this case with `instanceof`. 

Additionally, change the error string in the `message` of the `DecodeError` to be more useful as determined by our users. 

I dig it

File: packages/superagent-wrapper/src/request.ts (hunk 4 above)
Comment (ericcrosson-bitgo) [reply packages/superagent-wrapper/src/request.ts:162]: https://github.com/BitGo/api-ts/pull/170

File: packages/superagent-wrapper/test/request.test.ts (hunk 2 above)
Comment (ericcrosson-bitgo) [reply packages/superagent-wrapper/test/request.test.ts:209]: https://github.com/BitGo/api-ts/pull/171

File: packages/superagent-wrapper/src/request.ts (hunk 3 above)
Comment (ericcrosson-bitgo) [reply packages/superagent-wrapper/src/request.ts:167]: hurm how would you want to handle this case, @bitgopatmcl? It's pretty gnarly

I was about to open a ticket to track and then I had the thought -- is it even possible for `res.original.body` to contain a circular reference here? Meaning, was `res.original.body` recently `JSON.parse`d? If so, then I suppose there's no cause for concern (though I'll leave a note indicating that we've considered the unsafe use of `JSON.stringify` and determined that we're in the clear)

Comment (bitgopatmcl) [reply packages/superagent-wrapper/src/request.ts:167]: What about if the server returns a non-`application/json` response? We should probably consider that case too.

**Comment by jonball4** [comment]
🍾  thank you guys

**Pull request merged**